restore/restore.go

```diff
@@ -213,3 +221,53 @@ func registerMetrics() {
 		prometheus.MustRegister(downloadErrors)
 	})
 }
+
+type downloadPlan struct {
+	files        map[string]digest.ForRestore
+	changedFiles map[string][]digest.ForRestore
+}
+
+func (dp *downloadPlan) addHost(prefix string, nodePlan plan.NodePlan) {
+	if dp.files == nil {
+		dp.files = make(map[string]digest.ForRestore)
+	}
+	for fileName, fileDigest := range nodePlan.Files {
+		if prefix != "" {
+			fileName = path.Join(prefix, fileName)
+		}
+		dp.files[fileName] = fileDigest
+	}
+	if len(nodePlan.ChangedFiles) > 0 {
+		if dp.changedFiles == nil {
+			dp.changedFiles = make(map[string][]digest.ForRestore)
+		}
+		for fileName, historyEntries := range nodePlan.ChangedFiles {
+			if prefix != "" {
+				fileName = path.Join(prefix, fileName)
+			}
+			historyForFile := dp.changedFiles[fileName]
+			for _, entry := range historyEntries {
+				historyForFile = append(historyForFile, entry.Digest)
+			}
+			dp.changedFiles[fileName] = historyForFile
+		}
+	}
+}
+
+func (dp *downloadPlan) includeChanged(prefix string) map[string]digest.ForRestore {
+	if len(dp.changedFiles) == 0 {
+		return dp.files
+	}
+
+	result := make(map[string]digest.ForRestore)
+	for fileName, fileDigest := range dp.files {
+		result[fileName] = fileDigest
+	}
+	for fileName, versions := range dp.changedFiles {
+		for versionId, versionDigest := range versions {
```

Comment:
We might save up some memory by ranging over `dp.changedFiles[fileName]` instead of ranging `versions`. And same above.

Reply:
`versions` is _not_ a copy of the contents of the slice. See: [slice internals](https://blog.golang.org/go-slices-usage-and-internals#TOC_4.)

(Also, I think `gc` is smart enough to avoid continuously re-looking up through the map if you were to write out `dp.changedFiles[fileName]`, but if it's not that's quite slow.)

Reply:
It will be a copy of `digest.ForRestore` in the previous range and a copy of the pointer of `[]HistoryEntry`. Just something to keep in mind as I've seen memory improvements when map ranging and avoiding full copy.